File: src/main/scala/laoflch/debezium/connector/informix/InformixConnectorTask.scala
/*
 * Copyright Debezium Authors.
 *
 * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 */
package laoflch.debezium.connector.informix

import io.debezium.config.{Configuration, Field}
import io.debezium.connector.base.ChangeEventQueue
import io.debezium.connector.common.BaseSourceTask
import io.debezium.pipeline.{ChangeEventSourceCoordinator, DataChangeEvent, ErrorHandler, EventDispatcher}
import io.debezium.pipeline.spi.{ChangeEventCreator, OffsetContext}
import io.debezium.relational.history.DatabaseHistory
import io.debezium.relational.{HistorizedRelationalDatabaseConnectorConfig, TableId}
import io.debezium.util.{Clock, SchemaNameAdjuster}
import org.apache.kafka.connect.errors.ConnectException
import org.apache.kafka.connect.source.SourceRecord
import org.slf4j.{Logger, LoggerFactory}

import java.sql.SQLException
import java.util
import java.util.Collections
import scala.jdk.CollectionConverters


/**
 * The main task executing streaming from Informix.
 * Responsible for lifecycle management the streaming code.
 *
 * @author laoflch Luo
 *
 */
object InformixConnectorTask {
  private val LOGGER = LoggerFactory.getLogger(classOf[InformixConnectorTask])
  private val CONTEXT_NAME = "informix-server-connector-task"

  class changeEventCreator() extends ChangeEventCreator {

    override def createDataChangeEvent(x: SourceRecord): DataChangeEvent = {
      new DataChangeEvent(x)
    }
  }
}

class InformixConnectorTask extends BaseSourceTask {

  private val LOGGER: Logger = LoggerFactory.getLogger(classOf[InformixConnectorTask])
  private var queue: ChangeEventQueue[DataChangeEvent] = null
  private var dataConnection: InformixConnection = null
  private var metadataConnection: InformixConnection = null
  private var schema: InformixDatabaseSchema = null

  override def version: String = Module.version

  override def start(config: Configuration): ChangeEventSourceCoordinator = {

    val connectorConfig = new InformixConnectorConfig(config)
    val topicSelector = InformixTopicSelector.defaultSelector(connectorConfig)
    val schemaNameAdjuster = SchemaNameAdjuster.create(InformixConnectorTask.LOGGER)

    val jdbcConfig = config.edit
      .withDefault("database.responseBuffering", "adaptive")
      .withDefault("database.fetchSize", 10000)
      .withDefault("informix.debug.file", "/tmp/debezium_informix_trace.out")
      .build
      .filter((x: String) => !(
        (x.startsWith(DatabaseHistory.CONFIGURATION_FIELD_PREFIX_STRING) ||
          x == HistorizedRelationalDatabaseConnectorConfig.DATABASE_HISTORY.name)))
      .subset("database.", true);

    this.dataConnection = new InformixConnection(jdbcConfig)
    this.metadataConnection = new InformixConnection(jdbcConfig)

    try {
      this.dataConnection.setAutoCommit(false)
    } catch {
      case e: SQLException =>
        e.printStackTrace()
        throw new ConnectException(e)
    }

    this.schema = new InformixDatabaseSchema(connectorConfig, schemaNameAdjuster, topicSelector, this.dataConnection)
    this.schema.initializeStorage()

    val previousOffset = getPreviousOffset(new (InformixOffsetContext.Loader)(connectorConfig))
    if (previousOffset != null)
      this.schema.recover(previousOffset)

    val taskContext = new InformixTaskContext(connectorConfig, this.schema)
    val clock = Clock.system
    // Set up the task record queue ...
    this.queue = new ChangeEventQueue.Builder[DataChangeEvent]()
      .pollInterval(connectorConfig.getPollInterval) // DEFAULT_POLL_INTERVAL_MILLIS = 500;
      .maxBatchSize(connectorConfig.getMaxBatchSize) // DEFAULT_MAX_BATCH_SIZE = 2048;
      .maxQueueSize(connectorConfig.getMaxQueueSize) // DEFAULT_MAX_QUEUE_SIZE = 8192;
      .loggingContextSupplier(() => taskContext.configureLoggingContext(InformixConnectorTask.CONTEXT_NAME))
      .build

    //createChangeEventQueueBuilder(connectorConfig,taskContext,InformixConnectorTask.CONTEXT_NAME)

    val errorHandler = new ErrorHandler(classOf[InformixConnector], connectorConfig.getLogicalName, this.queue)
    val metadataProvider = new InformixEventMetadataProvider

    val dispatcher = new EventDispatcher[TableId](connectorConfig,
      topicSelector,
      this.schema,
      this.queue,
      connectorConfig.getTableFilters.dataCollectionFilter,
      new InformixConnectorTask.changeEventCreator(),
      metadataProvider)

    val coordinator = new ChangeEventSourceCoordinator(previousOffset,
      errorHandler,
      classOf[InformixConnector],
      connectorConfig,
      new InformixChangeEventSourceFactory(connectorConfig, this.dataConnection, this.metadataConnection,
        errorHandler, dispatcher, clock, this.schema),
      dispatcher,
      this.schema)

    coordinator.start(taskContext, this.queue, metadataProvider)
    coordinator
  }

  /**
   * Loads the connector's persistent offset (if present) via the given loader.
   */
  override protected def getPreviousOffset(loader: OffsetContext.Loader): OffsetContext = {
    val partition = loader.getPartition
    val previousOffset = context.offsetStorageReader.offsets(Collections.singleton(partition)).get(partition)

    //previousOffset.put("",this.dataConnection.getCDCEngine())
    if (previousOffset != null) {
      val offsetContext = loader.load(previousOffset)
      InformixConnectorTask.LOGGER.info("Found previous offset {}", offsetContext)
      offsetContext
    }
    else null
  }

  @throws[InterruptedException]
  override def doPoll: util.List[SourceRecord] = {
    //import scala.collection.JavaConversions._
    CollectionConverters.BufferHasAsJava(CollectionConverters.ListHasAsScala(this.queue.poll).asScala.map[SourceRecord]((x: DataChangeEvent) => x.getRecord)).asJava
  }

  override def doStop(): Unit = {
    try if (dataConnection != null) {
      dataConnection.close()
    }
    catch {
      case e: SQLException =>
        InformixConnectorTask.LOGGER.error("Exception while closing JDBC connection", e)
    }
    try if (metadataConnection != null) {
      metadataConnection.close()
    }
    catch {
      case e: SQLException =>
        InformixConnectorTask.LOGGER.error("Exception while closing JDBC metadata connection", e)
    }
    if (schema != null) {
      schema.close()
    }
  }

  override protected def getAllConfigurationFields: java.lang.Iterable[Field] = {
    InformixConnectorConfig.ALL_FIELDS.asInstanceOf[java.lang.Iterable[Field]]
  }
}

File: src/main/java/io/debezium/connector/informix/InformixConnectorTask.java
/*
 * Copyright Debezium Authors.
 *
 * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
 */
package io.debezium.connector.informix;

import java.sql.SQLException;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.apache.kafka.connect.errors.ConnectException;
import org.apache.kafka.connect.source.SourceRecord;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import io.debezium.bean.StandardBeanNames;
import io.debezium.config.CommonConnectorConfig;
import io.debezium.config.Configuration;
import io.debezium.config.Field;
import io.debezium.connector.base.ChangeEventQueue;
import io.debezium.connector.common.BaseSourceTask;
import io.debezium.document.DocumentReader;
import io.debezium.jdbc.DefaultMainConnectionProvidingConnectionFactory;
import io.debezium.jdbc.MainConnectionProvidingConnectionFactory;
import io.debezium.pipeline.ChangeEventSourceCoordinator;
import io.debezium.pipeline.DataChangeEvent;
import io.debezium.pipeline.ErrorHandler;
import io.debezium.pipeline.EventDispatcher;
import io.debezium.pipeline.metrics.DefaultChangeEventSourceMetricsFactory;
import io.debezium.pipeline.notification.NotificationService;
import io.debezium.pipeline.signal.SignalProcessor;
import io.debezium.pipeline.spi.Offsets;
import io.debezium.relational.TableId;
import io.debezium.schema.SchemaNameAdjuster;
import io.debezium.snapshot.SnapshotterService;
import io.debezium.spi.topic.TopicNamingStrategy;
import io.debezium.util.Clock;

/**
 * The main task executing streaming from Informix.
 * Responsible for lifecycle management the streaming code.
 *
 * @author Jiri Pechanec, Laoflch Luo, Lars M Johansson
 *
 */
public class InformixConnectorTask extends BaseSourceTask<InformixPartition, InformixOffsetContext> {

    private static final Logger LOGGER = LoggerFactory.getLogger(InformixConnectorTask.class);

    private static final String CONTEXT_NAME = "informix-server-connector-task";

    private volatile InformixTaskContext taskContext;
    private volatile ChangeEventQueue<DataChangeEvent> queue;
    private volatile InformixConnection dataConnection;
    private volatile InformixConnection cdcConnection;
    private volatile ErrorHandler errorHandler;
    private volatile InformixDatabaseSchema schema;

    @Override
    public String version() {
        return Module.version();
    }

    @Override
    protected ChangeEventSourceCoordinator<InformixPartition, InformixOffsetContext> start(Configuration config) {
        final InformixConnectorConfig connectorConfig = new InformixConnectorConfig(config);
        final TopicNamingStrategy<TableId> topicNamingStrategy = connectorConfig.getTopicNamingStrategy(CommonConnectorConfig.TOPIC_NAMING_STRATEGY);
        final SchemaNameAdjuster schemaNameAdjuster = connectorConfig.schemaNameAdjuster();

        MainConnectionProvidingConnectionFactory<InformixConnection> connectionFactory = new DefaultMainConnectionProvidingConnectionFactory<>(
                () -> new InformixConnection(connectorConfig.getJdbcConfig()));
        MainConnectionProvidingConnectionFactory<InformixConnection> cdcConnectionFactory = new DefaultMainConnectionProvidingConnectionFactory<>(
                () -> new InformixConnection(connectorConfig.getCdcJdbcConfig()));
        dataConnection = connectionFactory.mainConnection();
        cdcConnection = cdcConnectionFactory.mainConnection();
        try {
            dataConnection.setAutoCommit(false);
        }
        catch (SQLException e) {
            throw new ConnectException(e);
        }

        final InformixValueConverters valueConverters = new InformixValueConverters(connectorConfig.getDecimalMode(), connectorConfig.getTemporalPrecisionMode(),
                connectorConfig.binaryHandlingMode());
        schema = new InformixDatabaseSchema(connectorConfig, topicNamingStrategy, valueConverters, schemaNameAdjuster, dataConnection);
        schema.initializeStorage();

        Offsets<InformixPartition, InformixOffsetContext> previousOffsets = getPreviousOffsets(new InformixPartition.Provider(connectorConfig),
                new InformixOffsetContext.Loader(connectorConfig));

        // Manual Bean Registration
        connectorConfig.getBeanRegistry().add(StandardBeanNames.CONFIGURATION, config);
        connectorConfig.getBeanRegistry().add(StandardBeanNames.CONNECTOR_CONFIG, connectorConfig);
        connectorConfig.getBeanRegistry().add(StandardBeanNames.DATABASE_SCHEMA, schema);
        connectorConfig.getBeanRegistry().add(StandardBeanNames.JDBC_CONNECTION, dataConnection);
        connectorConfig.getBeanRegistry().add(StandardBeanNames.VALUE_CONVERTER, valueConverters);
        connectorConfig.getBeanRegistry().add(StandardBeanNames.OFFSETS, previousOffsets);

        // Service providers
        registerServiceProviders(connectorConfig.getServiceRegistry());

        final InformixPartition partition = previousOffsets.getTheOnlyPartition();
        final InformixOffsetContext previousOffset = previousOffsets.getTheOnlyOffset();

        final SnapshotterService snapshotterService = connectorConfig.getServiceRegistry().tryGetService(SnapshotterService.class);

        validateAndLoadSchemaHistory(connectorConfig, dataConnection::validateLogPosition, previousOffsets, schema,
                snapshotterService.getSnapshotter());

        taskContext = new InformixTaskContext(connectorConfig, schema);

        final Clock clock = Clock.system();

        // Set up the task record queue ...
        this.queue = new ChangeEventQueue.Builder<DataChangeEvent>()
                .pollInterval(connectorConfig.getPollInterval())
                .maxBatchSize(connectorConfig.getMaxBatchSize())
                .maxQueueSize(connectorConfig.getMaxQueueSize())
                .loggingContextSupplier(() -> taskContext.configureLoggingContext(CONTEXT_NAME))
                .build();

        errorHandler = new InformixErrorHandler(connectorConfig, queue, errorHandler);

        final InformixEventMetadataProvider metadataProvider = new InformixEventMetadataProvider();

        final SignalProcessor<InformixPartition, InformixOffsetContext> signalProcessor = new SignalProcessor<>(
                InformixConnector.class,
                connectorConfig,
                Map.of(),
                getAvailableSignalChannels(),
                DocumentReader.defaultReader(),
                previousOffsets);

        final EventDispatcher<InformixPartition, TableId> dispatcher = new EventDispatcher<>(
                connectorConfig,
                topicNamingStrategy,
                schema,
                queue,
                connectorConfig.getTableFilters().dataCollectionFilter(),
                DataChangeEvent::new,
                null,
                connectorConfig.createHeartbeat(topicNamingStrategy, schemaNameAdjuster, null, null),
                schemaNameAdjuster,
                new InformixTransactionMonitor(
                        connectorConfig,
                        metadataProvider,
                        schemaNameAdjuster,
                        (record) -> {
                            queue.enqueue(new DataChangeEvent(record));
                        },
                        topicNamingStrategy.transactionTopic()),
                signalProcessor);

        final NotificationService<InformixPartition, InformixOffsetContext> notificationService = new NotificationService<>(
                getNotificationChannels(),
                connectorConfig,
                InformixSchemaFactory.get(),
                dispatcher::enqueueNotification);

        final ChangeEventSourceCoordinator<InformixPartition, InformixOffsetContext> coordinator = new ChangeEventSourceCoordinator<>(
                previousOffsets,
                errorHandler,
                InformixConnector.class,
                connectorConfig,
                new InformixChangeEventSourceFactory(connectorConfig, connectionFactory, cdcConnectionFactory, errorHandler, dispatcher, clock, schema,
                        snapshotterService),
                new DefaultChangeEventSourceMetricsFactory<>(),
                dispatcher, schema,
                signalProcessor,
                notificationService,
                snapshotterService);

        coordinator.start(taskContext, this.queue, metadataProvider);

        return coordinator;
    }

    @Override
    protected List<SourceRecord> doPoll() throws InterruptedException {

        return queue.poll().stream()
                .map(DataChangeEvent::getRecord)
                .collect(Collectors.toList());
    }

    @Override
    protected void doStop() {
        try {
            if (dataConnection != null) {
                // Informix may have an active in-progress transaction associated with the connection and if so,
                // it will throw an exception during shutdown because the active transaction exists. This
                // is meant to help avoid this by rolling back the current active transaction, if exists.
                if (dataConnection.isConnected()) {
                    try {
                        dataConnection.rollback();
                    }
                    catch (SQLException e) {
                        // ignore
                    }
                }
                dataConnection.close();
            }
        }
        catch (SQLException e) {
            LOGGER.error("Exception while closing JDBC connection", e);
        }

        try {
            if (cdcConnection != null) {
                cdcConnection.close();
            }
        }
        catch (SQLException e) {
            LOGGER.error("Exception while closing CDC JDBC connection", e);
        }

        if (schema != null) {
            schema.close();
        }
    }

    @Override
    protected Iterable<Field> getAllConfigurationFields() {
        return InformixConnectorConfig.ALL_FIELDS;
    }
}
